lib/util/pthread_create_blocking.c

```diff
@@ -0,0 +1,279 @@
+	int * rc_init_p;	/* Non-zero indicates an init failure. */
+	int * rc_sync_p;	/* 1 if condition variable signalling failed. */
+};
+
+/* Set the "initialized" flag and signal the condition variable. */
+static int
+set_initialized(struct wrapped_cookie * U)
+{
+	int rc;
+
+	/* Lock mutex. */
+	if ((rc = pthread_mutex_lock(U->mutex_p))) {
+		warn0("pthread_mutex_lock: %s", strerror(rc));
+		goto err0;
+	}
+
+	/* Set the "initialized" flag and signal the condition variable. */
+	*U->initialized_p = 1;
+	if ((rc = pthread_cond_signal(U->cond_p))) {
+		warn0("pthread_cond_signal: %s", strerror(rc));
+		goto err1;
+	}
+
+	/* Unlock mutex. */
+	if ((rc = pthread_mutex_unlock(U->mutex_p))) {
+		warn0("pthread_mutex_unlock: %s", strerror(rc));
+		goto err0;
+	}
+
+	/* Success! */
+	return (0);
+
+err1:
+	if ((rc = pthread_mutex_unlock(U->mutex_p)))
+		warn0("pthread_mutex_unlock: %s", strerror(rc));
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/**
+ * pthread_cleanup_{push,pop}() can be implemented as either functions or as
+ * macros, and must appear as a pair in the same lexical scope.  As such,
+ * we cannot have a construct such as
+ *     if (clean_func != NULL)
+ *             pthread_cleanup_push(clean_func, user_cookie);
+ * Instead, if the user has passed a NULL for clean_func, we set it to this
+ * "no-op" function.
+ */
+static void
+cleanup_fake(void * user_cookie)
+{
+
+	(void)user_cookie; /* UNUSED */
+}
+
+static void *
+wrapped_thread(void * cookie)
+{
+	struct wrapped_cookie * U = cookie;
+	void * (*run_routine)(void *) = U->run_routine;
+	void (*cleanup_routine)(void *) = U->cleanup_routine;
+	void * user_cookie = U->user_cookie;
+	void * rc;
+
+	/* Run the user's init function, if provided. */
+	if (U->init_routine) {
+		if ((*U->rc_init_p = U->init_routine(user_cookie)) != 0) {
+			if (set_initialized(U) != 0)
+				*U->rc_sync_p = -1;
+			return (NULL);
+		}
+	}
+
+	/*
+	 * Set up the cleanup function.  If none was specified, use a "no-op"
+	 * function (see the comment for cleanup_fake()),
+	 */
+	if (cleanup_routine == NULL)
+		cleanup_routine = cleanup_fake;
+	pthread_cleanup_push(cleanup_routine, user_cookie);
+
+	/*
+	 * Indicate that the main thread can continue (which will free U), and
+	 * run the desired routine. */
+	if (set_initialized(U) != 0) {
+		*U->rc_sync_p = -1;
+		rc = NULL;
+	} else {
+		/* Run the desired routine. */
+		rc = run_routine(user_cookie);
+	}
+
+	/* Clean up. */
+	pthread_cleanup_pop(1);
+
+	/* Done. */
+	return (rc);
+}
+
+/* Wait up 1 second, or until the condition variable is signalled. */
+static int
+cond_timedwait(struct wrapped_cookie * U)
+{
+	struct timespec tp = {0};
+	struct timeval tv;
+	int rc;
+
+	/* 1 second later than the current time. */
+	if (monoclock_get(&tv)) {
+		warnp("monoclock_get");
+		goto err0;
+	}
+	tp.tv_sec = tv.tv_sec;
+	tp.tv_nsec = 1000 * tv.tv_usec;
+
+	/* Wait for up to 1 second. */
+	if ((rc = pthread_cond_timedwait(U->cond_p, U->mutex_p, &tp)) != 0) {
```

Comment:
I'm inclined to just say that if `pthread` functions are breaking, we're probably going to end up with deadlocks sooner or later so we might as well just accept it.  If you're really worried, mark the mutex as robust so that `pthread_cond_wait` will return if the spawned thread dies.